### PR TITLE
Add a neural style transfer model with AdaIN

### PR DESCRIPTION
**Pull request by chunxiangzheng**

### CODE REVIEW

**Comment by LukeWood** [comment]
Notes from Luke/Jake 1:1:

get_adain_encoder(), get_adain_decoder() returns a tuple of encoder and decoder
Internally, we can use AdaIn as a custom layer
AdaIn(include_rescaling=..., weights=”mine.h5”) -> this is a functional model that just calls
encoder()
AdaIn() layer
decoder()
Internal that creates an AdaIn identical model from an encoder and decoder
model = produceAdaIn(encoder, decoder) -> functional model
Then we can call model.save(“mymodel.h5”)
examples/models/training/style_transfer/adain.py -> define the model trainer the same way we have defined it before

File: keras_cv/models/adain.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (LukeWood) [inline keras_cv/models/adain.py:1]: API discussion:

- we want two functions, one that creates an encoder, decoder, and functional model end to end.  return all 3
- another function that calls the first, drops the first two, and loads weights into the functional model.

File: keras_cv/models/adain.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                padding="same",
+                activation="sigmoid",
+            ),
+        ]
+    )
+    return decoder    
+
+
+def get_loss_net(image_size=(None, None)):
+    vgg19 = keras.applications.VGG19(
```
Comment (LukeWood) [inline keras_cv/models/adain.py:66]: lets do the same thing with include_rescaling here

File: keras_cv/models/adain.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        """
+        style, content = inputs
+        content_mean, content_std = get_mean_std(content)
+        style_mean, style_std = get_mean_std(style)
+        t = style_std * (content - content_mean) / content_std + style_mean
+        return t
+
+
+class AdaINModel(keras.Model):
+    def __init__(self, image_size=(None, None), style_weight=4.0, **kwargs):
```
Comment (LukeWood) [inline keras_cv/models/adain.py:109]: then you'll add `include_rescaling` as a required argument after `self`

File: keras_cv/models/adain.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        content_input = layers.Input([*self.image_size, 3])
+        content_encoded = self.encoder(content_input)
+
+        adain_feature_map = self.adain_layer((style_encoded, content_encoded))
+        content_decoded = self.decoder(adain_feature_map)
+        
+        return keras.Model([style_input, content_input], content_decoded, name="adain_style_transfer")
+
+
+    def compile(self, optimizer, loss_fn):
```
Comment (LukeWood) [inline keras_cv/models/adain.py:130]: you can drop the custom method here and just rely on `self.optimizer` and `self.compiled_loss` instead of `self.loss_fn`

File: keras_cv/models/adain.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        super().compile()
+        self.optimizer = optimizer
+        self.loss_fn = loss_fn
+        self.style_loss_tracker = keras.metrics.Mean(name="style_loss")
```
Comment (LukeWood) [inline keras_cv/models/adain.py:134]: lets initialize these in the constructor

File: keras_cv/models/adain.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            The AdaIN feature map.
+        """
+        style, content = inputs
+        content_mean, content_std = get_mean_std(content)
+        style_mean, style_std = get_mean_std(style)
+        t = style_std * (content - content_mean) / content_std + style_mean
+        return t
+
+
+class AdaINModel(keras.Model):
```
Comment (LukeWood) [inline keras_cv/models/adain.py:108]: lets name this class `AdaInTrainer`

File: keras_cv/models/adain.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        style, content = inputs
+        # Encode the style and content image.
+        style_encoded = self.encoder(style)
+        content_encoded = self.encoder(content)
+
+        # Compute the AdaIN target feature maps.
+        t = self.adain_layer((style_encoded, content_encoded))
+
+        # Generate the neural style transferred image.
+        return self.decoder(t), t
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/adain.py:148]: Let's actually do this in the decoder - that way we don't have to think about it, and the decoder goes from the learned space `R_n` => image in range 0, 255.

So in the production of the decoder, lets add a multiply by 0, 255 at the end of the network and clip to 0, 255.  This will probably required you to migrate to a functional model - otherwise SavedModel will break for the Sequential.

you'll basically just iterate the loops you have defined, then add:

```
x = tf.clip_by_value(x * 255., 0., 255.0)
```

File: keras_cv/models/adain.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        with tf.GradientTape() as tape:
+            reconstructed_image, t = self(inputs)
+            # Compute the losses.
+            reconstructed_vgg_features = self.loss_net(reconstructed_image)
+            style_vgg_features = self.loss_net(style)
+            loss_content = self.loss_fn(t, reconstructed_vgg_features[-1])
+            for inp, out in zip(style_vgg_features, reconstructed_vgg_features):
+                mean_inp, std_inp = get_mean_std(inp)
+                mean_out, std_out = get_mean_std(out)
+                loss_style += self.loss_fn(mean_inp, mean_out) + self.loss_fn(
```
Comment (LukeWood) [inline keras_cv/models/adain.py:167]: you'll be able to update this to `self.compiled_loss(mean_inp, mean_out) + self.compiled_loss(std_inp, std_out)`

File: keras_cv/models/adain.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+
+def get_mean_std(x):
+    """Computes the mean and standard deviation of a tensor."""
+    axes = [1, 2]
+    mean, variance = tf.nn.moments(x, axes=axes, keepdims=True)
+    standard_deviation = tf.sqrt(variance + 1e-5)
+    return mean, standard_deviation
+
+
```
Comment (LukeWood) [inline keras_cv/models/adain.py:94]: @tf.keras.utils.register_keras_serializable(package="keras_cv")

File: keras_cv/models/adain.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+def get_mean_std(x):
+    """Computes the mean and standard deviation of a tensor."""
+    axes = [1, 2]
+    mean, variance = tf.nn.moments(x, axes=axes, keepdims=True)
+    standard_deviation = tf.sqrt(variance + 1e-5)
+    return mean, standard_deviation
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class AdaIn(layers.Layer):
+    def __init__(self, name='AdaIn', **kwargs):
```
Comment (LukeWood) [inline keras_cv/models/adain.py:93]: feel free to remove the default name

File: keras_cv/models/adain.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        super(AdaIn, self).__init__(name=name, **kwargs)
```
Comment (LukeWood) [inline keras_cv/models/adain.py:94]: super() is sufficient in python3:

```
        super().__init__(**kwargs)
```

File: keras_cv/models/adain.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+def get_mean_std(x):
+    """Computes the mean and standard deviation of a tensor."""
+    axes = [1, 2]
+    mean, variance = tf.nn.moments(x, axes=axes, keepdims=True)
+    standard_deviation = tf.sqrt(variance + 1e-5)
+    return mean, standard_deviation
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class AdaIn(layers.Layer):
```
Comment (LukeWood) [inline keras_cv/models/adain.py:92]: lets move this into its own file

File: keras_cv/models/adain.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (LukeWood) [inline keras_cv/models/adain.py:1]: lets move adain into it's own directory.

Lets do:

keras_Cv/models/generative/adain/

**Review by LukeWood — CHANGES_REQUESTED** [review]

File: keras_cv/models/adain.py (hunk 12 above)
Comment (LukeWood) [inline keras_cv/models/adain.py:92]: lets rename to `AdaptiveInstanceNormalization`

File: keras_cv/models/adain.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+    x = layers.Conv2D(filters=128, **config)(x)
+    x = layers.UpSampling2D()(x)
+    x = layers.Conv2D(filters=64, **config)(x)
+    x = layers.Conv2D(
+                filters=3,
+                kernel_size=3,
+                strides=1,
+                padding="same",
+                activation="sigmoid")(x)
+    outputs = tf.clip_by_value(x * 255., 0., 255.0)
```
Comment (LukeWood) [inline keras_cv/models/adain.py:62]: we can remove this and just include it at inference time.

**Comment by LukeWood** [comment]
As far as the API is concerned; lets wrap the inference logic in a top level class that loads the proper weights to the models, and implements inference in a `style_transfer()` method.  You can see an example of this here: https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/generative/stable_diffusion/stable_diffusion.py#L41

File: examples/training/style_transfer/ada_in/training.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+        plt.close()
+
+trainer.compile(optimizer=optimizer, loss=loss_fn)
+history = trainer.fit(
+    train_ds,
+    epochs=flags.EPOCHS,
+    steps_per_epoch=50,
+    validation_data=val_ds,
+    validation_steps=50,
+    callbacks=[TrainMonitor()],
```
Comment (LukeWood) [inline examples/training/style_transfer/ada_in/training.py:189]: lets also add a model checkpoint callback

File: examples/training/style_transfer/ada_in/training.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+test_ds = (
+    tf.data.Dataset.zip((test_style_ds, test_content_ds))
+    .shuffle(BATCH_SIZE * 2)
+    .apply(tf.data.experimental.dense_to_ragged_batch(batch_size=BATCH_SIZE))
+    .prefetch(AUTOTUNE)
+)
+
+trainer = AdaInTrainer(include_rescaling=True, image_size=IMAGE_SIZE)
+optimizer = keras.optimizers.Adam(learning_rate=1e-5)
```
Comment (LukeWood) [inline examples/training/style_transfer/ada_in/training.py:150]: another place to look as far as the bad results go is here;  the learning rate might be too low or too high.

**Comment by tanzhenyu** [comment]
Thanks for the PR! I'm not sure about the context here. It's not aligned with our immediate roadmap, so if this is important, please go ahead and address the comments so we can merge it. Or we can close it for now, and if this becomes needed we can re-open it later.

**Comment by ianstenbit** [comment]
Hey @chunxiangzheng -- where did we land on this PR?
If you're interested in finishing it up, please feel free to address comments and re-request a review from Luke.

I'm going to close this PR for now, but just re-open it if you'd like to keep it going!
